File: app/src/main/java/me/wsj/fengyun/view/fragment/WeatherFragment.java
package me.wsj.fengyun.view.fragment;

import android.annotation.SuppressLint;
import android.os.Build;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.view.ViewGroup;

import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.qweather.sdk.bean.WarningBean;
import com.qweather.sdk.bean.air.AirNowBean;
import com.qweather.sdk.bean.weather.WeatherDailyBean;
import com.qweather.sdk.bean.weather.WeatherHourlyBean;
import com.qweather.sdk.bean.weather.WeatherNowBean;

import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;

import java.util.ArrayList;
import java.util.List;

import me.wsj.fengyun.R;
import me.wsj.fengyun.adapter.ForecastAdapter;
import me.wsj.fengyun.databinding.FragmentWeatherBinding;
import me.wsj.fengyun.databinding.LayoutTodayDetailBinding;
import me.wsj.fengyun.presenters.WeatherInterface;
import me.wsj.fengyun.presenters.impl.WeatherImpl;
import me.wsj.fengyun.utils.ContentUtil;
import me.wsj.fengyun.utils.IconUtils;
import me.wsj.fengyun.utils.TransUnitUtil;
import me.wsj.fengyun.view.base.BaseFragment;
import me.wsj.fengyun.widget.horizonview.ScrollWatched;
import me.wsj.fengyun.widget.horizonview.ScrollWatcher;
import per.wsj.commonlib.utils.LogUtil;

public class WeatherFragment extends BaseFragment<FragmentWeatherBinding> implements WeatherInterface {
    private static final String PARAM_CITY_ID = "param_city_id";
    List<ScrollWatcher> watcherList = new ArrayList<>();

    private ScrollWatched watched;

    private String currentTime;
    private String sunrise;
    private String sunset;
    private String moonRise;
    private String moonSet;
    private boolean hasAni = false;

    private String todayMaxTmp;
    private String todayMinTmp;
    private WeatherDailyBean weatherForecastBean;
    private WeatherHourlyBean weatherHourlyBean;
    private String nowTmp;
    private String mCityId;

    private String condCode;

    private ForecastAdapter forecastAdapter;

    private LayoutTodayDetailBinding todayDetailBinding;

    private WeatherViewModel viewModel;

    public static WeatherFragment newInstance(String cityId) {
        WeatherFragment fragment = new WeatherFragment();
        Bundle args = new Bundle();
        args.putString(PARAM_CITY_ID, cityId);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mCityId = getArguments().getString(PARAM_CITY_ID);
        }
    }


    @Override
    protected FragmentWeatherBinding bindView() {
        return FragmentWeatherBinding.inflate(getLayoutInflater());
    }

    @Override
    public void initView(View view) {
        viewModel = new ViewModelProvider(getActivity()).get(WeatherViewModel.class);
        todayDetailBinding = LayoutTodayDetailBinding.bind(mBinding.getRoot());

        getCurrentTime();

        mBinding.horizontalScrollView.setToday24HourView(mBinding.hourly);

        //横向滚动监听
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            mBinding.horizontalScrollView.setOnScrollChangeListener(new View.OnScrollChangeListener() {
                @Override
                public void onScrollChange(View v, int scrollX, int scrollY, int oldScrollX, int oldScrollY) {
                    watched.notifyWatcher(scrollX);
                }
            });
        }

        mBinding.swipeLayout.setOnRefreshListener(() -> initData(mCityId));
    }

    @Override
    public void initEvent() {
        watched = new ScrollWatched() {
            @Override
            public void addWatcher(ScrollWatcher watcher) {
                watcherList.add(watcher);
            }

            @Override
            public void removeWatcher(ScrollWatcher watcher) {
                watcherList.remove(watcher);
            }

            @Override
            public void notifyWatcher(int x) {
                for (ScrollWatcher watcher : watcherList) {
                    watcher.update(x);
                }
            }
        };

        watched.addWatcher(mBinding.hourly);
    }

    @Override
    public void loadData() {
        initData(mCityId);

        if (!hasAni && !TextUtils.isEmpty(sunrise) && !TextUtils.isEmpty(sunset) && !TextUtils.isEmpty(moonRise) && !TextUtils.isEmpty(moonSet)) {
            setViewTime();
        }

        if (!TextUtils.isEmpty(sunrise) && !TextUtils.isEmpty(sunset) && !TextUtils.isEmpty(moonRise) && !TextUtils.isEmpty(moonSet)) {
            setViewTime();
        }

        getWeatherForecast(weatherForecastBean);
    }

    private void initData(String cityId) {
        WeatherImpl weatherImpl = new WeatherImpl(this.getActivity(), this);
        weatherImpl.getWeatherHourly(cityId);
        weatherImpl.getAirForecast(cityId);
        weatherImpl.getAirNow(cityId);
        weatherImpl.getWarning(cityId);
        weatherImpl.getWeatherForecast(cityId);
        weatherImpl.getWeatherNow(cityId);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onResume() {
        super.onResume();

        LogUtil.e("-----------------onResume------------------" + mCityId);
        if (condCode != null) {
            viewModel.setCondCode(condCode);
            LogUtil.e("-----------------ch bg------------------ ch bg" + mCityId);
        }

        if (!hasAni && !TextUtils.isEmpty(sunrise) && !TextUtils.isEmpty(sunset) && !TextUtils.isEmpty(moonRise) && !TextUtils.isEmpty(moonSet)) {
            setViewTime();
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        LogUtil.e("-----------------onDetach------------------" + mCityId);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        LogUtil.e("-----------------onDestroy------------------" + mCityId);
    }

    @SuppressLint("SetTextI18n")
    public void changeUnit() {
        if (todayDetailBinding.tvMaxTmp != null) {
            if (ContentUtil.APP_SETTING_UNIT.equals("hua")) {
                LogUtil.e("当前城市1：" + condCode);
                todayDetailBinding.tvMaxTmp.setText(TransUnitUtil.getF(todayMaxTmp) + "°F");
                todayDetailBinding.tvMinTmp.setText(TransUnitUtil.getF(todayMinTmp) + "°F");
                mBinding.tvTodayTmp.setText(TransUnitUtil.getF(nowTmp) + "°F");
            } else {
                LogUtil.e("当前城市2：" + condCode);
                todayDetailBinding.tvMaxTmp.setText(todayMaxTmp + "°C");
                todayDetailBinding.tvMinTmp.setText(todayMinTmp + "°C");
                mBinding.tvTodayTmp.setText(nowTmp + "°C");
            }
        }
        getWeatherHourly(weatherHourlyBean);
        getWeatherForecast(weatherForecastBean);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void getWeatherNow(WeatherNowBean bean) {
        if (bean != null && bean.getNow() != null) {
            WeatherNowBean.NowBaseBean now = bean.getNow();
            String rain = now.getPrecip();
            String hum = now.getHumidity();
            String pres = now.getPressure();
            String vis = now.getVis();
            String windDir = now.getWindDir();
            String windSc = now.getWindScale();
            String condTxt = now.getText();
            condCode = now.getIcon();
            viewModel.setCondCode(condCode);

            nowTmp = now.getTemp();
            mBinding.tvTodayCond.setText(condTxt);
            mBinding.tvTodayTmp.setText(nowTmp + "°C");
            if (ContentUtil.APP_SETTING_UNIT.equals("hua")) {
                mBinding.tvTodayTmp.setText(TransUnitUtil.getF(nowTmp) + "°F");
            }
            todayDetailBinding.tvTodayRain.setText(rain + "mm");
            todayDetailBinding.tvTodayPressure.setText(pres + "HPA");
            todayDetailBinding.tvTodayHum.setText(hum + "%");
            todayDetailBinding.tvTodayVisible.setText(vis + "KM");
            todayDetailBinding.tvWindDir.setText(windDir);
            todayDetailBinding.tvWindSc.setText(windSc + "级");
            DateTime nowTime = DateTime.now();
            int hourOfDay = nowTime.getHourOfDay();
            if (hourOfDay > 6 && hourOfDay < 19) {
                mBinding.ivBack.setImageResource(IconUtils.getDayBack(getContext(), condCode));
            } else {
                mBinding.ivBack.setImageResource(IconUtils.getNightBack(getContext(), condCode));
            }

            mBinding.swipeLayout.setRefreshing(false);
        }
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void getWeatherForecast(WeatherDailyBean bean) {
        if (bean != null && bean.getDaily() != null) {
            weatherForecastBean = bean;
            getCurrentTime();

            List<WeatherDailyBean.DailyBean> daily_forecast = bean.getDaily();

            WeatherDailyBean.DailyBean forecastBase = daily_forecast.get(0);
            String condCodeD = forecastBase.getIconDay();
            String condCodeN = forecastBase.getIconNight();
            String tmpMin = forecastBase.getTempMin();
            String tmpMax = forecastBase.getTempMax();
            sunrise = forecastBase.getSunrise();
            sunset = forecastBase.getSunset();
            moonRise = forecastBase.getMoonRise();
            moonSet = forecastBase.getMoonSet();
            todayDetailBinding.sunView.setTimes(sunrise, sunset, currentTime);
            todayDetailBinding.moonView.setTimes(moonRise, moonSet, currentTime);
            todayMaxTmp = tmpMax;
            todayMinTmp = tmpMin;
            todayDetailBinding.tvMaxTmp.setText(tmpMax + "°");
            todayDetailBinding.tvMinTmp.setText(tmpMin + "°");
            todayDetailBinding.ivTodayDay.setImageResource(IconUtils.getDayIconDark(getContext(), condCodeD));
            todayDetailBinding.ivTodayNight.setImageResource(IconUtils.getNightIconDark(getContext(), condCodeN));

            if (forecastAdapter == null) {
                forecastAdapter = new ForecastAdapter(getActivity(), daily_forecast);
                mBinding.rvForecast.setAdapter(forecastAdapter);
                LinearLayoutManager forecastManager = new LinearLayoutManager(getActivity());
                forecastManager.setOrientation(LinearLayoutManager.VERTICAL);
                mBinding.rvForecast.setLayoutManager(forecastManager);
            } else {
                forecastAdapter.refreshData(getActivity(), daily_forecast);
            }

        }
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void getWarning(WarningBean.WarningBeanBase alarmBase) {
        if (alarmBase != null) {
            mBinding.tvTodayAlarm.setVisibility(View.VISIBLE);
            String level = alarmBase.getLevel();
            String type = alarmBase.getType();
            mBinding.tvTodayAlarm.setText(type + "预警");

            if (!TextUtils.isEmpty(level)) {
                switch (level) {
                    case "蓝色":
                    case "Blue":
                        mBinding.tvTodayAlarm.setBackground(getResources().getDrawable(R.drawable.shape_blue_alarm));
                        mBinding.tvTodayAlarm.setTextColor(getResources().getColor(R.color.white));
                        break;
                    case "黄色":
                    case "Yellow":
                        mBinding.tvTodayAlarm.setBackground(getResources().getDrawable(R.drawable.shape_yellow_alarm));
                        mBinding.tvTodayAlarm.setTextColor(getResources().getColor(R.color.white));
                        break;
                    case "橙色":
                    case "Orange":
                        mBinding.tvTodayAlarm.setBackground(getResources().getDrawable(R.drawable.shape_orange_alarm));
                        mBinding.tvTodayAlarm.setTextColor(getResources().getColor(R.color.white));
                        break;
                    case "红色":
                    case "Red":
                        mBinding.tvTodayAlarm.setBackground(getResources().getDrawable(R.drawable.shape_red_alarm));
                        mBinding.tvTodayAlarm.setTextColor(getResources().getColor(R.color.white));
                        break;
                    case "白色":
                    case "White":
                        mBinding.tvTodayAlarm.setBackground(getResources().getDrawable(R.drawable.shape_white_alarm));
                        mBinding.tvTodayAlarm.setTextColor(getResources().getColor(R.color.black));
                        break;
                }
            }
        } else {
            mBinding.tvTodayAlarm.setVisibility(View.GONE);
        }
    }

    @Override
    public void getAirNow(AirNowBean bean) {
        if (bean != null && bean.getNow() != null) {
            todayDetailBinding.ivLine2.setVisibility(View.VISIBLE);
            todayDetailBinding.gridAir.setVisibility(View.VISIBLE);
            todayDetailBinding.rvAir.setVisibility(View.VISIBLE);
            todayDetailBinding.airTitle.setVisibility(View.VISIBLE);
            AirNowBean.NowBean airNowCity = bean.getNow();
            String qlty = airNowCity.getCategory();
            String aqi = airNowCity.getAqi();
            String pm25 = airNowCity.getPm2p5();
            String pm10 = airNowCity.getPm10();
            String so2 = airNowCity.getSo2();
            String no2 = airNowCity.getNo2();
            String co = airNowCity.getCo();
            String o3 = airNowCity.getO3();
            todayDetailBinding.tvAir.setText(qlty);
            todayDetailBinding.tvAirNum.setText(aqi);
            todayDetailBinding.tvTodayPm25.setText(pm25);
            todayDetailBinding.tvTodayPm10.setText(pm10);
            todayDetailBinding.tvTodaySo2.setText(so2);
            todayDetailBinding.tvTodayNo2.setText(no2);
            todayDetailBinding.tvTodayCo.setText(co);
            todayDetailBinding.tvTodayO3.setText(o3);
            todayDetailBinding.rvAir.setBackground(viewModel.getAirBackground(aqi));
        } else {
            todayDetailBinding.ivLine2.setVisibility(View.GONE);
            todayDetailBinding.gridAir.setVisibility(View.GONE);
            todayDetailBinding.rvAir.setVisibility(View.GONE);
            todayDetailBinding.airTitle.setVisibility(View.GONE);
        }
    }


    @SuppressLint("SetTextI18n")
    @Override
    public void getWeatherHourly(WeatherHourlyBean bean) {
        if (bean != null && bean.getHourly() != null) {
            weatherHourlyBean = bean;
            List<WeatherHourlyBean.HourlyBean> hourlyWeatherList = bean.getHourly();
            List<WeatherHourlyBean.HourlyBean> data = new ArrayList<>();
            if (hourlyWeatherList.size() > 23) {
                for (int i = 0; i < 24; i++) {
                    data.add(hourlyWeatherList.get(i));
                    String condCode = data.get(i).getIcon();
                    String time = data.get(i).getFxTime();
                    time = time.substring(time.length() - 11, time.length() - 9);
                    int hourNow = Integer.parseInt(time);
                    if (hourNow >= 6 && hourNow <= 19) {
                        data.get(i).setIcon(condCode + "d");
                    } else {
                        data.get(i).setIcon(condCode + "n");
                    }
                }
            } else {
                for (int i = 0; i < hourlyWeatherList.size(); i++) {
                    data.add(hourlyWeatherList.get(i));
                    String condCode = data.get(i).getIcon();
                    String time = data.get(i).getFxTime();
                    time = time.substring(time.length() - 11, time.length() - 9);
                    int hourNow = Integer.parseInt(time);
                    if (hourNow >= 6 && hourNow <= 19) {
                        data.get(i).setIcon(condCode + "d");
                    } else {
                        data.get(i).setIcon(condCode + "n");
                    }
                }
            }

            int minTmp = Integer.parseInt(data.get(0).getTemp());
            int maxTmp = minTmp;
            for (int i = 0; i < data.size(); i++) {
                int tmp = Integer.parseInt(data.get(i).getTemp());
                minTmp = Math.min(tmp, minTmp);
                maxTmp = Math.max(tmp, maxTmp);
            }
            //设置当天的最高最低温度
            mBinding.hourly.setHighestTemp(maxTmp);
            mBinding.hourly.setLowestTemp(minTmp);
            if (maxTmp == minTmp) {
                mBinding.hourly.setLowestTemp(minTmp - 1);
            }
            mBinding.hourly.initData(data);
            mBinding.tvLineMaxTmp.setText(maxTmp + "°");
            mBinding.tvLineMinTmp.setText(minTmp + "°");
            if (ContentUtil.APP_SETTING_UNIT.equals("hua")) {
                mBinding.tvLineMaxTmp.setText(TransUnitUtil.getF(String.valueOf(maxTmp)) + "°");
                mBinding.tvLineMinTmp.setText(TransUnitUtil.getF(String.valueOf(minTmp)) + "°");
            }
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (mBinding.getRoot() != null && mBinding.getRoot().getParent() != null) {
            ((ViewGroup) mBinding.getRoot().getParent()).removeView(mBinding.getRoot());
        }
    }

    /**
     * 设置view的时间
     */
    private void setViewTime() {
        getCurrentTime();

        todayDetailBinding.sunView.setTimes(sunrise, sunset, currentTime);
        todayDetailBinding.moonView.setTimes(moonRise, moonSet, currentTime);
        hasAni = true;
    }

    private void getCurrentTime() {
        DateTime now = DateTime.now(DateTimeZone.UTC);
        float a = Float.parseFloat("+8.0");
        float minute = a * 60;
        now = now.plusMinutes(((int) minute));
        currentTime = now.toString("HH:mm");
    }

}

File: app/src/main/java/me/wsj/fengyun/view/fragment/WeatherFragment.kt
package me.wsj.fengyun.view.fragment

import android.annotation.SuppressLint
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import me.wsj.fengyun.R
import me.wsj.fengyun.adapter.ForecastAdapter
import me.wsj.fengyun.bean.*
import me.wsj.fengyun.databinding.FragmentWeatherBinding
import me.wsj.fengyun.databinding.LayoutAirQualityBinding
import me.wsj.fengyun.databinding.LayoutSunMoonBinding
import me.wsj.fengyun.databinding.LayoutTodayDetailBinding
import me.wsj.lib.extension.notEmpty
import me.wsj.lib.extension.toastCenter
import me.wsj.fengyun.utils.ContentUtil
import me.wsj.lib.utils.IconUtils
import me.wsj.fengyun.utils.Lunar
import me.wsj.fengyun.utils.WeatherUtil
import me.wsj.fengyun.view.activity.vm.MainViewModel
import me.wsj.fengyun.view.base.BaseVmFragment
import me.wsj.fengyun.view.base.LoadState
import me.wsj.fengyun.widget.horizonview.ScrollWatched
import me.wsj.fengyun.widget.horizonview.ScrollWatcher
import org.joda.time.DateTime
import org.joda.time.DateTimeZone
import per.wsj.commonlib.utils.LogUtil
import per.wsj.commonlib.utils.Typefaces
import java.util.*

private const val PARAM_CITY_ID = "param_city_id"


class WeatherFragment : BaseVmFragment<FragmentWeatherBinding, WeatherViewModel>() {

    private lateinit var mCityId: String

    private val watcherList by lazy { ArrayList<ScrollWatcher>() }

    private lateinit var watched: ScrollWatched

    private var currentTime: String? = null
    private var sunrise: String? = null
    private var sunset: String? = null
    private var moonRise: String? = null
    private var moonSet: String? = null
    private var hasAni = false

    private var todayMaxTmp: String? = null
    private var todayMinTmp: String? = null

    private var nowTmp: String? = null

    private var condCode: String? = null

    private var forecastAdapter: ForecastAdapter? = null

    private var todayDetailBinding: LayoutTodayDetailBinding? = null

    private lateinit var airQualityBinding: LayoutAirQualityBinding

    private lateinit var sunMoonBinding: LayoutSunMoonBinding

    private lateinit var mainViewModel: MainViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            mCityId = it.getString(PARAM_CITY_ID).toString()
        }
    }

    companion object {
        @JvmStatic
        fun newInstance(param1: String) =
            WeatherFragment().apply {
                arguments = Bundle().apply {
                    putString(PARAM_CITY_ID, param1)
                }
            }
    }

    override fun onStart() {
        super.onStart()
        viewModel.loadCache(mCityId)

        Calendar.getInstance().apply {
            val day = get(Calendar.DAY_OF_MONTH)
            mBinding.tvDate.text =
                (get(Calendar.MONTH) + 1).toString() + "月" + day + "日 农历" +
                        Lunar(this).toString()
        }
    }

    override fun onResume() {
        super.onResume()
        condCode?.let {
            LogUtil.e("onResume() set cond code: $condCode")
            mainViewModel.setCondCode(it)
        }

        setViewTime()
    }

    override fun bindView() = FragmentWeatherBinding.inflate(layoutInflater)

    override fun initView(view: View?) {
        // must use activity
        mainViewModel = ViewModelProvider(requireActivity()).get(MainViewModel::class.java)

        todayDetailBinding = LayoutTodayDetailBinding.bind(mBinding.root)

        sunMoonBinding = LayoutSunMoonBinding.bind(mBinding.root)

        airQualityBinding = LayoutAirQualityBinding.bind(mBinding.root)

        mBinding.horizontalScrollView.setToday24HourView(mBinding.hourly)

        watched = object : ScrollWatched {
            override fun addWatcher(watcher: ScrollWatcher) {
                watcherList.add(watcher)
            }

            override fun removeWatcher(watcher: ScrollWatcher) {
                watcherList.remove(watcher)
            }

            override fun notifyWatcher(x: Int) {
                for (watcher in watcherList) {
                    watcher.update(x)
                }
            }
        }

        watched.addWatcher(mBinding.hourly)

        //横向滚动监听
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            mBinding.horizontalScrollView.setOnScrollChangeListener { _, scrollX, _, _, _ ->
                watched.notifyWatcher(scrollX)
            }
        }

        // 设置字体
        mBinding.tvTodayTmp.typeface = Typefaces.get(requireContext(), "widget_clock.ttf")
    }

    override fun initEvent() {
        mBinding.swipeLayout.setOnRefreshListener { loadData() }

        viewModel.weatherNow.observe(this) {
            showWeatherNow(it)
        }

        viewModel.warnings.observe(this) {
            showWarnings(it)
        }

        viewModel.airNow.observe(this) {
            showAirNow(it)
        }

        viewModel.forecast.observe(this) {
            showForecast(it)
        }
        viewModel.hourly.observe(this) {
            showHourly(it)
        }

        viewModel.loadState.observe(this) {
            when (it) {
                is LoadState.Start -> {
                    mBinding.swipeLayout.isRefreshing = true
                }
                is LoadState.Error -> {

                }
                is LoadState.Finish -> {
                    if (viewModel.isStopped()) {
                        mBinding.swipeLayout.isRefreshing = false
                    }
                }
            }
        }
    }

    override fun loadData() {
        viewModel.loadData(mCityId)
    }

    @SuppressLint("SetTextI18n")
    fun showWeatherNow(now: Now) {
        condCode = now.icon
        nowTmp = now.temp
        lifecycleScope.launchWhenResumed {
            LogUtil.e("showWeatherNow() set cond code: $condCode")
            mainViewModel.setCondCode(now.icon)
        }
        mBinding.tvTodayCond.text = now.text
        mBinding.tvTodayTmp.text = now.temp
        mBinding.tvUnit.visibility = View.VISIBLE

        if (ContentUtil.APP_SETTING_UNIT == "hua") {
            mBinding.tvTodayTmp.text = WeatherUtil.getF(now.temp).toString() + "°F"
        }
        todayDetailBinding!!.tvTodayRain.text = now.precip + "mm"
        todayDetailBinding!!.tvTodayPressure.text = now.pressure + "HPA"
        todayDetailBinding!!.tvTodayHum.text = "${now.humidity}%"
        todayDetailBinding!!.tvTodayVisible.text = now.vis + "KM"
        todayDetailBinding!!.tvWindDir.text = now.windDir
        todayDetailBinding!!.tvWindSc.text = now.windScale + "级"
    }

    /**
     * 三天预报
     */
    private fun showForecast(dailyForecast: List<Daily>) {
        getCurrentTime()
        val forecastBase = dailyForecast[0]
        val condCodeD = forecastBase.iconDay
        val condCodeN = forecastBase.iconNight
        todayMinTmp = forecastBase.tempMin
        todayMaxTmp = forecastBase.tempMax
        sunrise = forecastBase.sunrise
        sunset = forecastBase.sunset
        moonRise = forecastBase.moonrise
        moonSet = forecastBase.moonset
        sunMoonBinding.sunView.setTimes(sunrise, sunset, currentTime)
        sunMoonBinding.moonView.setTimes(moonRise, moonSet, currentTime)
        todayDetailBinding!!.tvMaxTmp.text = "$todayMaxTmp°"
        todayDetailBinding!!.tvMinTmp.text = "$todayMinTmp°"
        todayDetailBinding!!.ivTodayDay.setImageResource(
            IconUtils.getDayIconDark(
                context,
                condCodeD
            )
        )
        todayDetailBinding!!.ivTodayNight.setImageResource(
            IconUtils.getNightIconDark(
                context,
                condCodeN
            )
        )
        if (forecastAdapter == null) {
            forecastAdapter = ForecastAdapter(activity, dailyForecast)
            mBinding.rvForecast.adapter = forecastAdapter
            val forecastManager = LinearLayoutManager(
                activity
            )
            forecastManager.orientation = LinearLayoutManager.VERTICAL
            mBinding.rvForecast.layoutManager = forecastManager
        } else {
            forecastAdapter!!.refreshData(activity, dailyForecast)
        }
    }

    /**
     * 空气质量
     */
    private fun showAirNow(airNow: Air) {
        airQualityBinding.gridAir.visibility = View.VISIBLE
        airQualityBinding.rvAir.visibility = View.VISIBLE
        airQualityBinding.airTitle.visibility = View.VISIBLE

        airQualityBinding.tvAir.text = airNow.category
        airQualityBinding.tvAirNum.text = airNow.aqi
        airQualityBinding.tvTodayPm25.text = airNow.pm2p5
        airQualityBinding.tvTodayPm10.text = airNow.pm10
        airQualityBinding.tvTodaySo2.text = airNow.so2
        airQualityBinding.tvTodayNo2.text = airNow.no2
        airQualityBinding.tvTodayCo.text = airNow.co
        airQualityBinding.tvTodayO3.text = airNow.o3
        airQualityBinding.rvAir.background =
            WeatherUtil.getAirBackground(requireContext(), airNow.aqi)
    }

    /**
     * 预警
     */
    private fun showWarnings(warnings: List<Warning>) {
        mBinding.alarmFlipper.visibility = View.VISIBLE
        mBinding.alarmFlipper.setInAnimation(requireContext(), R.anim.bottom_in)
        mBinding.alarmFlipper.setOutAnimation(requireContext(), R.anim.top_out)
        mBinding.alarmFlipper.flipInterval = 3000
        for (warning in warnings) {
            val level: String = warning.level
            val tip = warning.typeName + level + "预警"
            val warningRes = WeatherUtil.getWarningRes(requireContext(), level)
            val textView: TextView = layoutInflater.inflate(R.layout.item_warning, null) as TextView
            textView.background = warningRes.first
            textView.text = tip
            textView.setOnClickListener {
                toastCenter(warning.text)
            }
            textView.setTextColor(warningRes.second)
            mBinding.alarmFlipper.addView(textView)
        }
        if (warnings.size > 1) {
            mBinding.alarmFlipper.startFlipping()
        }
    }

    /**
     * 逐小时天气
     */
    private fun showHourly(hourlyWeatherList: List<Hourly>) {
        val data: MutableList<Hourly> = ArrayList()
        if (hourlyWeatherList.size > 23) {
            for (i in 0..23) {
                data.add(hourlyWeatherList[i])
                val condCode = data[i].icon
                var time = data[i].fxTime
                time = time.substring(time.length - 11, time.length - 9)
                val hourNow = time.toInt()
                if (hourNow in 6..19) {
                    data[i].icon = condCode + "d"
                } else {
                    data[i].icon = condCode + "n"
                }
            }
        } else {
            for (i in hourlyWeatherList.indices) {
                data.add(hourlyWeatherList[i])
                val condCode = data[i].icon
                var time = data[i].fxTime
                time = time.substring(time.length - 11, time.length - 9)
                val hourNow = time.toInt()
                if (hourNow in 6..19) {
                    data[i].icon = condCode + "d"
                } else {
                    data[i].icon = condCode + "n"
                }
            }
        }
        var minTmp = data[0].temp.toInt()
        var maxTmp = minTmp
        for (i in data.indices) {
            val tmp = data[i].temp.toInt()
            minTmp = Math.min(tmp, minTmp)
            maxTmp = Math.max(tmp, maxTmp)
        }
        //设置当天的最高最低温度
        mBinding.hourly.setHighestTemp(maxTmp)
        mBinding.hourly.setLowestTemp(minTmp)
        if (maxTmp == minTmp) {
            mBinding.hourly.setLowestTemp(minTmp - 1)
        }
        mBinding.hourly.initData(data)
        mBinding.tvLineMaxTmp.text = "$maxTmp°"
        mBinding.tvLineMinTmp.text = "$minTmp°"
        if (ContentUtil.APP_SETTING_UNIT == "hua") {
            mBinding.tvLineMaxTmp.text = WeatherUtil.getF(maxTmp.toString()).toString() + "°"
            mBinding.tvLineMinTmp.text = WeatherUtil.getF(minTmp.toString()).toString() + "°"
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        if (mBinding.root.parent != null) {
            (mBinding.root.parent as ViewGroup).removeView(mBinding.root)
        }
    }

    /**
     * 设置view的时间
     */
    private fun setViewTime() {
        if (!hasAni && sunrise.notEmpty() && sunset.notEmpty() && moonRise.notEmpty() && moonSet.notEmpty()) {
            getCurrentTime()
            sunMoonBinding.sunView.setTimes(sunrise, sunset, currentTime)
            sunMoonBinding.moonView.setTimes(moonRise, moonSet, currentTime)
            hasAni = true
        }
    }

    private fun getCurrentTime() {
        var now = DateTime.now(DateTimeZone.UTC)
        val a = "+8.0".toFloat()
        val minute = a * 60
        now = now.plusMinutes(minute.toInt())
        currentTime = now.toString("HH:mm")
    }

    fun changeUnit() {
        if (ContentUtil.APP_SETTING_UNIT == "hua") {
            LogUtil.e("当前城市1：$condCode")
            todayDetailBinding!!.tvMaxTmp.text =
                WeatherUtil.getF(todayMaxTmp!!).toString() + "°F"
            todayDetailBinding!!.tvMinTmp.text =
                WeatherUtil.getF(todayMinTmp!!).toString() + "°F"
            mBinding.tvTodayTmp.text = WeatherUtil.getF(nowTmp!!).toString() + "°F"
        } else {
            LogUtil.e("当前城市2：$condCode")
            todayDetailBinding!!.tvMaxTmp.text = "$todayMaxTmp°C"
            todayDetailBinding!!.tvMinTmp.text = "$todayMinTmp°C"
            mBinding.tvTodayTmp.text = "$nowTmp°C"
        }
//        getWeatherHourly(weatherHourlyBean)
//        getWeatherForecast(weatherForecastBean)
    }
}